{
    "assignee": null,
    "assignees": [],
    "body": "The following code crashes Cython 0.25.2. It compiles fine on 0.25.1. This is the whole file.\r\n\r\n```\r\n    def fn():\r\n        value = {}\r\n        for k, v in value.items():\r\n            value=k\r\n```\r\n\r\nThe following variant also causes a crash (again in 0.25.2 but not 0.25.1):\r\n\r\n```\r\n    def fn():\r\n        value = {}\r\n        for k, value in value.items():\r\n            pass\r\n```\r\n\r\nThe thing to notice about both these snippets is the variable `value`, which is used to define the iterator of the `for` loop, is assigned to in the loop. In the first example it is assigned to in the loop body, and in the second example it is assigned to in the `for` loop target list. Admittedly, both are bad code style.\r\n\r\nIf `value.items()` is changed to `value`, or if the iteration is put in global scope rather than in a function, then the crash doesn't happen. The `-2` and `-3` switches make no difference.\r\n\r\nHere is the full console output:\r\n\r\n```\r\nError compiling Cython file:\r\n------------------------------------------------------------\r\n...\r\ndef fn():\r\n    value = {}\r\n    for k, value in value.items():\r\n                              ^\r\n------------------------------------------------------------\r\n\r\nloop.py:7:31: Compiler crash in AnalyseExpressionsTransform\r\n\r\nModuleNode.body = StatListNode(loop.py:5:0)\r\nStatListNode.stats[0] = DefNode(loop.py:5:0,\r\n    is_cyfunction = True,\r\n    modifiers = [...]/0,\r\n    name = 'fn',\r\n    py_wrapper_required = True,\r\n    reqd_kw_flags_cname = '0',\r\n    used = True)\r\nFile 'Nodes.py', line 430, in analyse_expressions: StatListNode(loop.py:6:4)\r\nFile 'Nodes.py', line 6178, in analyse_expressions: ForInStatNode(loop.py:7:4)\r\nFile 'ExprNodes.py', line 519, in analyse_expressions: NextNode(loop.py:7:20,\r\n    result_is_used = True,\r\n    use_managed_ref = True)\r\nFile 'ExprNodes.py', line 2733, in analyse_types: NextNode(loop.py:7:20,\r\n    result_is_used = True,\r\n    use_managed_ref = True)\r\nFile 'ExprNodes.py', line 2730, in infer_type: NextNode(loop.py:7:20,\r\n    result_is_used = True,\r\n    use_managed_ref = True)\r\nFile 'ExprNodes.py', line 3310, in infer_type: IndexNode(loop.py:7:20,\r\n    is_subscript = True,\r\n    result_is_used = True,\r\n    use_managed_ref = True)\r\nFile 'ExprNodes.py', line 4959, in infer_type: SimpleCallNode(loop.py:7:31,\r\n    analysed = True,\r\n    is_temp = 1,\r\n    result_is_used = True,\r\n    use_managed_ref = True)\r\n\r\nCompiler crash traceback from this point on:\r\n  File \"/home/arthur/envs/default/lib/python3.5/site-packages/Cython/Compiler/ExprNodes.py\", line 4959, in infer_type\r\n    arg_types = [arg.infer_type(env) for arg in self.args]\r\nTypeError: 'NoneType' object is not iterable\r\n```",
    "closed_at": null,
    "comment_data": [],
    "comments": 0,
    "comments_url": "https://api.github.com/repos/cython/cython/issues/1640/comments",
    "created_at": "2017-03-22T16:36:01Z",
    "event_data": [],
    "events_url": "https://api.github.com/repos/cython/cython/issues/1640/events",
    "html_url": "https://github.com/cython/cython/issues/1640",
    "id": 216130527,
    "labels": [],
    "labels_url": "https://api.github.com/repos/cython/cython/issues/1640/labels{/name}",
    "locked": false,
    "milestone": null,
    "number": 1640,
    "repository_url": "https://api.github.com/repos/cython/cython",
    "state": "open",
    "title": "Compiler crash in AnalyseExpressionsTransform: TypeError: 'NoneType' object is not iterable",
    "updated_at": "2017-03-22T16:36:01Z",
    "url": "https://api.github.com/repos/cython/cython/issues/1640",
    "user": {
        "avatar_url": "https://avatars2.githubusercontent.com/u/20324099?v=3",
        "events_url": "https://api.github.com/users/arthur-tacca/events{/privacy}",
        "followers_url": "https://api.github.com/users/arthur-tacca/followers",
        "following_url": "https://api.github.com/users/arthur-tacca/following{/other_user}",
        "gists_url": "https://api.github.com/users/arthur-tacca/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/arthur-tacca",
        "id": 20324099,
        "login": "arthur-tacca",
        "organizations_url": "https://api.github.com/users/arthur-tacca/orgs",
        "received_events_url": "https://api.github.com/users/arthur-tacca/received_events",
        "repos_url": "https://api.github.com/users/arthur-tacca/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/arthur-tacca/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/arthur-tacca/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/arthur-tacca"
    }
}